### PR TITLE
feat: add support for basic auth, insecure tls and headers

### PR DESCRIPTION
**Pull request by ricardomaraschini**
## Description

add suport for basic auth, insecure tls and headers to the download package. this PR is part of an [adr](https://github.com/k0sproject/k0s/blob/main/docs/architecture/adr-001-autopilot-oci-basic-auth-support.md) implementation that i plan to introduce gradually.

Fixes # (issue)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

## How Has This Been Tested?

- [x] Manual test
- [x] Auto test added

## Checklist:

- [x] My code follows the style [guidelines](https://github.com/k0sproject/k0s/blob/main/docs/contributors/overview.md) of this project 
- [x] My commit messages are [signed-off](https://github.com/k0sproject/k0s/blob/main/docs/contributors/github_workflow.md)
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules
- [x] I have checked my code and corrected any misspellings

### CODE REVIEW

File: internal/http/download.go [hunk 1]
```diff
@@ -36,23 +37,35 @@ type DownloadOption func(*downloadOptions)
...
 		opt(&opts)
 	}
 
-	// Prepare the client and the request.
-	client := http.Client{
-		Transport: &http.Transport{
-			// This is a one-shot HTTP client which should release resources immediately.
-			DisableKeepAlives: true,
-		},
+	transport := http.DefaultTransport.(*http.Transport).Clone()
```
Comment (twz123) [inline internal/http/download.go:46]: It was a deliberate decision not to use the default transport here, as it defines a bunch of timeouts that make configuration difficult. I think for this simple use case the staleness timeout is all that's needed.

The only thing that's probably good to set here is `ProxyFromEnvironment`. I'd even say setting this here should actually be backported.

File: internal/http/download_test.go [hunk 2]
```diff
@@ -153,7 +205,17 @@ func startFakeDownloadServer(t *testing.T, handler http.Handler) string {
...
+			serverError <- server.Serve(listener)
+			return
+		}
+
+		cert, key := generateSelfSignedCert(t)
+		t.Cleanup(func() {
+			os.Remove(key)
+			os.Remove(cert)
+		})
+		serverError <- server.ServeTLS(listener, cert, key)
```
Comment (twz123) [inline internal/http/download_test.go:218]: There's also a shorter solution for this in [joinclient_test.go](https://github.com/k0sproject/k0s/blob/v1.31.1%2Bk0s.0/pkg/token/joinclient_test.go#L147-L177). No need to write the certs to disk, so there's no need for cleanups.

File: internal/http/download_test.go [hunk 3]
```diff
@@ -143,7 +150,52 @@ func TestDownload_RedirectLoop(t *testing.T) {
...
+	assert.ErrorContains(t, err, "bad status: 401 Unauthorized")
+	err = internalhttp.Download(context.TODO(), baseURL, io.Discard, internalhttp.WithBasicAuth("myuser", "mypassword"))
+	assert.NoError(t, err)
+}
+
+func TestDownload_InsecureSkipTLSVerify(t *testing.T) {
+	baseURL := startFakeDownloadServer(t, true, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		_, err := w.Write([]byte("content"))
+		require.NoError(t, err)
+		w.WriteHeader(http.StatusOK)
```
Comment (twz123) [inline internal/http/download_test.go:170]: We can also just leave the handler body empty, as the content is discarded later on, and the status defaults to OK anyways.

(The call to WriteHeader has no effect anyways, as it's happening after the first write.)

```suggestion
```

File: internal/http/download.go (hunk 1 above)
Comment (ricardomaraschini) [reply internal/http/download.go:46]: https://github.com/k0sproject/k0s/pull/5064

File: internal/http/download_test.go (hunk 3 above)
Comment (ricardomaraschini) [reply internal/http/download_test.go]: That makes sense. I have incorporated the proposed changes on my branch.

File: internal/http/download_test.go (hunk 2 above)
Comment (ricardomaraschini) [reply internal/http/download_test.go]: Way better. This is fixed here.

**Review by twz123 — APPROVED** [review]

**Pull request merged**